lib/Optimizer/GraphOptimizer/NodeSplitting.cpp

```diff
@@ -0,0 +1,1561 @@
+    }
+    if (!bigChunksFirst_ && (idx == 0)) {
+      chunkSize = chunkSizeSmall;
+    }
+    chunkSizes[idx] = chunkSize;
+  }
+  return chunkSizes;
+}
+
+std::vector<dim_t> SplitNodeByChunkSizes::splitAlongDim(size_t dim,
+                                                        dim_t dimSize) const {
+  size_t dimIdx = getSplitDimIdx(dim);
+  std::vector<dim_t> chunkSizes = chunkSizes_[dimIdx];
+  size_t numChunks = chunkSizes.size();
+  CHECK((1 <= numChunks) && (numChunks <= dimSize))
+      << "SplitNodeByChunkSizes: Invalid number of sizes '" << numChunks
+      << "' for splitting a dimension with size '" << dimSize << "'!";
+  for (const auto &chunkSize : chunkSizes) {
+    CHECK_GT(chunkSize, 0)
+        << "SplitNodeByChunkSizes: Chunk size 0 is not allowed!";
+  }
+  return chunkSizes;
+}
+
+std::vector<dim_t> SplitNodeByChunkWeights::splitAlongDim(size_t dim,
+                                                          dim_t dimSize) const {
+  size_t dimIdx = getSplitDimIdx(dim);
+  const std::vector<float> &chunkWeights = chunkWeights_[dimIdx];
+  dim_t numChunks = chunkWeights.size();
+  CHECK((1 <= numChunks) && (numChunks <= dimSize))
+      << "SplitNodeByChunkWeights: Invalid number of weights '" << numChunks
+      << "' for splitting a dimension with size '" << dimSize << "'!";
+
+  // Verify that all the weights are positive and compute the weights sum.
+  float chunkWeightsSum = 0;
+  for (const auto &weight : chunkWeights) {
+    CHECK_GT(weight, 0.f) << "SplitNodeByChunkWeights: Chunk weight '" << weight
+                          << "' invalid! Should be strictly positive!";
+    chunkWeightsSum += weight;
+  }
+
+  // Compute individual chunk sizes such that each chunk gets at least one unit
+  // (empty chunks are NOT allowed). The total number of units is distributed
+  // such that the error between the given chunk weights and the actual chunk
+  // weights is minimized.
+  std::vector<dim_t> chunkSizes(numChunks, 1);
+  dim_t unitsRem = dimSize - numChunks;
+  while (unitsRem > 0) {
+    // Find chunk with maximum weight error.
+    float weightErrMax = std::numeric_limits<float>::lowest();
+    size_t weightErrMaxIdx = 0;
+    for (size_t idx = 0; idx < numChunks; ++idx) {
+      float weightVal =
+          float(chunkSizes[idx]) / float(dimSize) * chunkWeightsSum;
+      // We use a signed error here to starve those chunks for which the actual
+      // weight surpassed the given weight.
+      float weightErr = chunkWeights[idx] - weightVal;
+      if (weightErr > weightErrMax) {
+        weightErrMax = weightErr;
+        weightErrMaxIdx = idx;
+      }
+    }
+    // Distribute unit.
+    chunkSizes[weightErrMaxIdx] += 1;
+    unitsRem--;
+  }
+  return chunkSizes;
+}
+
+/// Utility function to split an array of slice ranges \p ranges along the given
+/// dimension \p dim using the split option \p splitOption.
+static std::vector<SliceRange>
+splitSliceRanges(const std::vector<SliceRange> &ranges, size_t dim,
```

Comment:
ArrayRef?

Reply:
Mutable vector so I left it `std::vector`.